typing/env.ml

```diff
@@ -1655,29 +1669,40 @@ let rec components_of_module_maker
             let final_decl = Subst.type_declaration prefixing_sub fresh_decl in
             Btype.set_row_name final_decl
               (Subst.type_path prefixing_sub (Path.Pident id));
-            let constructors =
-              List.map snd
-                (Datarepr.constructors_of_type ~current_unit:(get_unit_name ())
-                   path final_decl)
+            let descrs = match decl.type_kind with
+            | Type_variant _ ->
+                let cstrs = List.map snd
+                  (Datarepr.constructors_of_type path final_decl
+                      ~current_unit:(get_unit_name ()))
+                in
+                List.iter
+                  (fun descr ->
+                    let cda = {
+                      cda_description = descr;
+                      cda_address = None }
+                    in
+                    c.comp_constrs <-
+                      add_to_tbl descr.cstr_name cda c.comp_constrs
+                  ) cstrs;
+               Type_variant cstrs
+            | Type_record (_, repr) ->
+                let lbls = List.map snd
+                  (Datarepr.labels_of_type path final_decl)
+                in
+                List.iter
+                  (fun descr ->
+                    c.comp_labels <-
+                      add_to_tbl descr.lbl_name descr c.comp_labels)
+                  lbls;
+                Type_record (lbls, repr)
+            | Type_abstract -> Type_abstract
+            | Type_open -> Type_open
```

Comment:
This is `bimap ~record:... ~variant:...`.

Reply:
Right. thanks! (I was looking at other parts of `env`.) Indeed one could define `bimap` (I would just call it `map`, given that there is no natural bias for a unimap) and use it here as a refactoring.